jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/ResourceMapping.kt

```diff
@@ -0,0 +1,86 @@
+import software.amazon.awssdk.services.cloudformation.CloudFormationClient
+import software.amazon.awssdk.services.cloudformation.model.ListStackResourcesRequest
+import software.amazon.awssdk.services.cloudformation.model.ListStacksRequest
+import software.amazon.awssdk.services.cloudformation.model.StackResourceSummary
+import software.amazon.awssdk.services.cloudformation.model.StackStatus
+import software.amazon.awssdk.services.cloudformation.model.StackSummary
+import software.aws.toolkits.jetbrains.core.AwsClientManager
+import java.util.Collections
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.CompletionStage
+import java.util.stream.Stream
+
+interface ResourceMapping {
+    fun listAllSummaries(logicalResource: String): Collection<StackResourceSummary>
+
+    fun guessPhysicalResourceId(logicalResource: String): String?
+
+    fun isEmpty(): Boolean
+
+    object EMPTY : ResourceMapping {
+        override fun listAllSummaries(logicalResource: String): Collection<StackResourceSummary> = Collections.emptyList()
+
+        override fun guessPhysicalResourceId(logicalResource: String): String? = null
+
+        override fun isEmpty(): Boolean = true
+    }
+}
+
+internal class ResourceMappingImpl : ResourceMapping {
+    private val logicalToSummary: MultiMap<String, StackResourceSummary> = MultiMap.createLinked()
+
+    override fun listAllSummaries(logicalResource: String): Collection<StackResourceSummary> = logicalToSummary.get(logicalResource)
+
+    override fun guessPhysicalResourceId(logicalResource: String): String? =
+            listAllSummaries(logicalResource).firstOrNull()?.physicalResourceId()
+
+    private fun registerResource(resourceSummary: StackResourceSummary) =
+            resourceSummary.logicalResourceId()?.let { logicalToSummary.putValue(it, resourceSummary) }
+
+    override fun isEmpty(): Boolean = logicalToSummary.isEmpty
+
+    companion object {
+        @JvmStatic
+        fun promiseMapping(project: Project): CompletionStage<ResourceMapping> {
+            val client: CloudFormationClient = AwsClientManager.getInstance(project).getClient()
+            return promiseMapping(client)
+        }
+
+        @JvmStatic
+        fun promiseMapping(client: CloudFormationClient): CompletionStage<ResourceMapping> {
+            val promise = CompletableFuture<ResourceMapping>()
+
+            ApplicationManager.getApplication().executeOnPooledThread {
+                val result = ResourceMappingImpl()
+                try {
+                    val stackRequest = ListStacksRequest.builder()
+                            .stackStatusFilters(StackStatus.CREATE_COMPLETE, StackStatus.UPDATE_COMPLETE)
+                            .build()
+                    client.listStacksPaginator(stackRequest).stream()
+                            .flatMap { it.stackSummaries().stream() }
+                            .flatMap { client.listResourcesForStack(it) }
```

Comment:
Our `AWS Resources` toolwindow is getting hammered by Cloudformations' low TPS - listResourcesForStack is 0.5 TPS with bursts to 10 - these calls get throttled extremely quickly. Instead of iterating over all the stacks I think we'll need to have the user select a stack they want to use to populate the values.

Nice thing about that is you can then get rid of the drop-down on the value itself - we'll just populate all values for the given reference off the stack selected (if we can find it).

The other thing I like about this approach is that we can (in the future) have a 'populate from remote lambda' option where the customer picks a lambda and we import the environment variables from there.